README
Digital Forensics Framework

1. Introduction

2. Installation
 2.1 Windows
 2.2 Linux
3. Usage
4. Support

1. Introduction

DFF is a simple but powerfull open source tool with a flexible module system
which will help you in your digital forensics works, including files recovery
due to error or crash, evidence research and analysis, etc. The source code is
written in C++ and Python, allowing performances and great extensibility.

Although dff is quite young, it already provides a robust architecture and
some handy modules. You could download and try it via the Dowload page. Source
code, Debian packages and even windows setup are available. Any contribution,
suggestion or remark is welcome ! (see 4. Support)

Nowadays computer forensic analysis tools are mainly large proprietary software
developed by some well-known companies.

Few free and open source tools offers the same type of fully integrated
software, most of them are implemented as stand alone tools. Although some
framework exists, they are not very user or developer friendly. That is why we
decided to develop this tool as a free and open source and multi-platform
framework.

This project follows three main goals :

- Modularity. In contrary to the monolithic model, the modular model is based on
  an a host and many modules. This modular conception presents two advantages :
  it permits to improve rapidly the software and to split easily tasks for
  developers

- Scriptability, it is obvious that the ability to be scripted gives more
  flexibility to a tool, but it enables automation and gives the possibility to
  extend features

- Genericity, the project tries to remain OS independent. We want to help people
  where they are ! Letting them choose any Operating System to use this
  software.


2. Installation

2.1 Windows

Python and Python QT have to be installed first.

Two packages are provided for Windows, one with those two dependencies add to DFF installer and an other with DFF installer alone.

Web-page where Python should be downloaded :
http://www.python.org/download/releases/2.7.1/.
Web-page where Python QT should be downloaded :
http://www.riverbankcomputing.co.uk/software/pyqt/download

DFF is provided with a Nullsoft installer. Users just have to launch it and
follow instructions to install DFF.


2.2 Linux

Using distribution package :

RPM and DEB packages are provided on http://www.digital-forensic.org. Graphical
helper from window manager can be used when double clicking on the package.

DEB installation from terminal :
#> dpkg -i dff-<version>.deb

RPM installation from terminal 
#> rpm -i dff-<version>.rpm

Compiling from sources :

A GZipped tarball is also provided.
Cmake and latest version of swig are needed (http://www.swig.org, developer have
to compile and install the latest version himself).
In the top-source tree type :
$> cmake -DINSTALL=TRUE .
It creates make files.
To build type :
$> make
Install :
#> make install
Console Run :
$> dff.py
Graphical Run :
$> dff.py -g

3. Usage

DFF reads a disk dump (for example from GNU 'dd' utils). Two user interfaces are
provided ; graphical and console. Command-line console also sits in graphical as
well.

Graphical :
Click on the 'File(s)' menu and select 'Open evidence file(s)', then add your
disk dump.
Right-click on the dump, in the 'Browser' tab, select 'Open with' and apply a
filesystem module ; select 'file system' and 'fat' for example.
Files appears in the 'Virtual File System' tab.
Many informations are provided under 'Task Manager', 'Output' and 'Errors' tabs.


Console :
Opening a local folder :
dff / > local --path /home/user/dumps --parent /

--path is the directory to open
--parent is a virtual node, first specify the root one : /

Applying fat module on a dump :
dff / > fat dumps/test.fat.dd

Listing nodes :
dff / > ls

Completion is provided using the <Tab> key. User can obtain help using :
dff / > man <command>


4. Support

Online chat is on an IRC channel : #digital-forensic on irc.freenode.net
network.

Main website : http://www.digital-forensic.org .

3 mailing lists are provided :
- User discussions about DFF : dff@digital-forensic.org, registration and
  posting freely available.
- Developers discussions about DFF : dff-devel@digital-forensic.org,
  registration and posting freely available.
- News about DFF releases and event : dff-announce@digital-forensic.org,
  registration freely available, low level traffic.

Archives of this mailing lists : http://lists.digital-forensic.org

A project manager exists at https://tracker.digital-forensic.org , ideas and
bug submited by e-mail will be reported on it.

Documentation sits on http://wiki.digital-forensic.org.



CHANGES
DFF (Digital Forensics Framework)

See the CHANGES.current file for changes in the current version.
See the RELEASENOTES file for a summary of changes in each release.

Version 1.3.0 (27 February 2013)
===============================

CHANGES after 10/04/11 are now split in corresponding category.

  * ui repository :
    -------------

Author: fba <fba@arxsys.fr>
Date:   Mon Feb 18 13:39:29 2013 +0100

    update of usage message

Author: fba <fba@arxsys.fr>
Date:   Thu Feb 14 10:59:01 2013 +0100

    fixes bad unicode handling in processus state

Author: fba <fba@arxsys.fr>
Date:   Tue Feb 12 18:15:13 2013 +0100

    update of french translation

Author: fba <fba@arxsys.fr>
Date:   Tue Feb 12 18:12:02 2013 +0100

    update of german and french translation

Author: sja <sja@digital-forensic.org>
Date:   Mon Feb 11 17:08:50 2013 +0100

    use inheritance for modules_path launcher

Author: fba <fba@arxsys.fr>
Date:   Mon Jan 14 19:21:22 2013 +0100

    * Update to new building macros

Author: fba <fba@arxsys.fr>
Date:   Mon Jan 14 13:46:24 2013 +0100

    * Changes in target creation

Author: sja <sja@digital-forensic.org>
Date:   Tue Dec 4 12:19:35 2012 +0100

    FIX for widget without titleName use widget name

Author: sja <sja@digital-forensic.org>
Date:   Mon Dec 3 19:27:49 2012 +0100

    FIX 'toolbar' translation for browser, now mainwindow use windowTitle as id not widget.name()

Author: fba <fba@arxsys.fr>
Date:   Mon Dec 3 11:45:45 2012 +0100

    Adds German translation provided by D. Schreiber

Author: sja <sja@digital-forensic.org>
Date:   Thu Nov 15 20:11:54 2012 +0100

    update translation for taskmanager

Author: sja <sja@digital-forensic.org>
Date:   Thu Nov 15 18:36:53 2012 +0100

    remove old indexation engine

Author: sja <sja@digital-forensic.org>
Date:   Thu Nov 15 16:36:30 2012 +0100

    update for translation

Author: sja <sja@digital-forensic.org>
Date:   Thu Nov 15 11:37:05 2012 +0100

    FIX translation loading

Author: sja <sja@digital-forensic.org>
Date:   Wed Nov 14 17:09:46 2012 +0100

    update french translation

Author: sja <sja@digital-forensic.org>
Date:   Thu Nov 8 17:48:32 2012 +0100

    fix import


  * modules repository :
    -------------------

Author: sja <sja@digital-forensic.org>
Date:   Mon Feb 25 11:42:30 2013 +0100

    FIX module local : typo in attribute name (orignal for original)

Author: sja <sja@digital-forensic.org>
Date:   Fri Feb 22 17:16:39 2013 +0100

    module exchange : rename __identifier to __id because it's keyword for VS ...

Author: sja <sja@digital-forensic.org>
Date:   Thu Feb 21 11:12:58 2013 +0100

    Change module/mfso name from pff to exchange

Author: sja <sja@digital-forensic.org>
Date:   Tue Feb 19 16:28:40 2013 +0100

    FIX #181 PFF fix bug with item attachments

Author: sja <sja@digital-forensic.org>
Date:   Mon Feb 18 19:30:57 2013 +0100

    FIX pff bug introducing a new structure ItemInfo to handle item type and recovery more easily and dynamically

Author: fba <fba@arxsys.fr>
Date:   Thu Feb 14 11:23:10 2013 +0100

    fixes issue with self.vnode which was still a variant

Author: sja <sja@digital-forensic.org>
Date:   Wed Feb 13 14:44:30 2013 +0100

    fix hexeditor add try/catch for open/close

Author: sja <sja@digital-forensic.org>
Date:   Mon Feb 11 17:41:09 2013 +0100

    pff_fix test

Author: sja <sja@digital-forensic.org>
Date:   Mon Feb 11 14:27:16 2013 +0100

    delete experimental will move it to parent directory

Author: sja <sja@digital-forensic.org>
Date:   Mon Feb 11 14:23:13 2013 +0100

    Module name, description or category was changed to be clearer

Author: fba <fba@arxsys.fr>
Date:   Thu Jan 31 13:07:55 2013 +0100

    * Carver GUI
    ** Adds checks and appropriate message box when:
    *** No items have been selected
    *** Input is a folder

Author: sja <sja@digital-forensic.org>
Date:   Tue Jan 29 12:15:14 2013 +0100

    FIX possible infinite loop on lnk modules applyied on carved or bad lnk

Author: fba <fba@arxsys.fr>
Date:   Wed Jan 16 14:42:43 2013 +0100

    *Build process Windows
    ** Changes *.dll with respective _DYN_LIBRARIES cmake variables

Author: jmo <jmo@arxsys.fr>
Date:   Tue Jan 15 15:13:40 2013 +0100

    Fixes #152 : Search from cursor error

Author: fba <fba@arxsys.fr>
Date:   Mon Jan 14 13:44:36 2013 +0100

    * Include pyrun.swg in all .i files
    * Changes CMakeLists.txt to support dff_cpp_module macro

Author: sja <sja@digital-forensic.org>
Date:   Mon Jan 7 18:13:16 2013 +0100

    Use new method to update state of the hash attributes

Author: fba <fba@arxsys.fr>
Date:   Mon Jan 7 15:06:36 2013 +0100

    adds a new config parameters for footer instead of using existing one. Made double free when deleting config

Author: fba <fba@arxsys.fr>
Date:   Fri Dec 14 16:04:26 2012 +0100

    * Extract
    ** adds better event handling for reporting failed extraction and renamed files

Author: fba <fba@arxsys.fr>
Date:   Fri Dec 7 18:39:21 2012 +0100

    * Modules regviewer
    ** Adds absolute "path" information and timestamp for each key

Author: cma <cma@digital-forensic.org>
Date:   Thu Dec 6 18:35:31 2012 +0100

    * NTFS improvements
    ** Adds filesystem parameters as attributes to the "NTFS" top-level node
    ** Adds the MFT cluster number as attribute for each file or directory
    ** Adds the first data cluster as attribute for each file
    ** Creates a "NTFS unallocated" node near the top-level module node
    *** Filled with unallocated area from the $Bitmap file
    *** Each node file is a continuous unallocated area named "starting cluster--ending cluster"
    *** This commit refs #167, need feedback.

Author: cma <cma@digital-forensic.org>
Date:   Mon Dec 3 17:52:04 2012 +0100

    * NTFS Fixes
    ** Infinite loop in directory parsing resulting in NTFS module never ends
    ** Random files detection on huge directory

Author: udgover <udgover@orion.(none)>
Date:   Wed Nov 28 06:53:34 2012 +0100


testsuite/local.py
#!/usr/bin/python 
# DFF -- An Open Source Digital Forensics Framework
# Copyright (C) 2009-2013 ArxSys
# This program is free software, distributed under the terms of
# the GNU General Public License Version 2. See the LICENSE file
# at the top of the source tree.
#  
# See http://www.digital-forensic.org for more information about this
# project. Please do not directly contact any of the maintainers of
# DFF for assistance; the project provides a web site, mailing lists
# and IRC channels for your use.
# 
# Author(s):
#  Christophe M. <cma@digital-forensic.org>
#

import unittest, sys, os, random
from dffunittest import DffUnittest

class LocalTests(DffUnittest):
    """ Tests on 'local' VFS driver
    We load several things, and validate basic functionality of the framework.
    Keep in mind that VFS loads are persistant from one test to another.

    TODO
    - seek
    - negative seek
    - open & close many files
    - load same file two times, open both, seek one, check both position are OK

    """
    
    def __init__(self, name='runTest'):
        unittest.TestCase.__init__(self, name)

# FIXME what about windows ?
    testFile = '/etc/passwd'

    def test01_LoadOneFile(self):
        """ #01 Load a file present on every Unix
        """

        # launch command
        self.ui.cmd('local --path ' + self.testFile + ' --parent /')
        
        # validate output from framework
        self.assertFalse(sys.stdout.getvalue())
        self.assertFalse(sys.stderr.getvalue())


    def test02_LsOneFile(self):
        """ #02 ls previously loaded file and validate size
        Check if --long (switch for size output) is right size of the file
        """
        
        self.ui.cmd('ls --long /')
        
        fileSize = str(os.stat(self.testFile).st_size)
        expectedStdout = '/' + os.path.basename(self.testFile) + '\t' + fileSize + '\n'

        self.assertEqual(sys.stdout.getvalue(), expectedStdout)
        self.assertFalse(sys.stderr.getvalue())
        

    def _itemsCount(self, path, onlyFiles = False):
        """ Return number of file in a given directory
        In case 'onlyFiles' is set, return only files count.
        """
        itemsCount = 0
        for root, dirs, files in os.walk(path):
            if onlyFiles:
                itemsCount += len(files)
            else:
                itemsCount += len(dirs) + len(files)
        return itemsCount

    def test03_LoadOneDirectory(self):
        """ #03 Load current working directory and validate number of files loaded
        Make sure number of nodes loaded is the same number as number of file
        stored in current working directory.
        """

        testDirectory = os.getcwd()

        self.ui.cmd('local --path ' + testDirectory + ' --parent /')

        # Expected output after local execution
# TODO Amount of loaded items is now in 'info -> local'
#        expectedItems = str(self._itemsCount(os.getcwd()))
#        expectedStdout = 'nodes created:\n' + expectedItems + '\nresult:\nno problem\n'
#        self.assertEqual(sys.stdout.getvalue(), expectedStdout)
        # In case of assertion failure above, hidden files exist in directory
        # loaded by this test

        # validate output from framework
        self.assertFalse(sys.stdout.getvalue())
        self.assertFalse(sys.stderr.getvalue())
        
    def test04_ReadFile(self):
        """ #04 Read 128 byte in previously loaded file
        Make sure content is equal on disk and in vfs
        """
        readAmount = 128
        
        with open(self.testFile, 'r') as f:
            expectedContent = f.read(128)

        checknode = self.vfs.getnode('/passwd')

        nodeFile = checknode.open()
        buff = nodeFile.read(128)
        self.assertEqual(expectedContent, buff)

    def _randomItem(self, path):
        """ Return a random filepath in a given directory
        """
        itemsCount = self._itemsCount(path, True)
        randomItemPosition = 0
        while randomItemPosition == 0 or randomItemPosition > itemsCount:
            randomItemPosition = int(random.random() * itemsCount)
        i = 0
        for root, dirs, files in os.walk(path):
            j = 0
            for j in range(len(files)):
                i += 1
                if i == randomItemPosition:
                    return root + '/' + files[j]


    def test05_RandomReadFile(self):
        """ #05 Randomly check 100 file content
        Read 'randomReadSize' from beginning of file and make sure
        same content is obtained from disk and from VFS.
        This test reads directly through the VFS object.
        """

        checkAmount = 100
        while checkAmount:
            # Obtain a random filepath from given directory
            randomFilepath = self._randomItem(os.getcwd())
            # Compute a random size to read
            randomReadSize = int(random.random() * os.stat(randomFilepath).st_size)
            # Open reference file
            with open(randomFilepath, 'r') as f:
                # Save reference data
                expectedContent = f.read(randomReadSize)
                # Set node path according to VFS format
                vfsNodePath = '/' + randomFilepath[len(os.getcwd()) - len(os.path.basename(os.getcwd())):]
                # Open node file
                nodeFile = self.vfs.getnode(vfsNodePath).open()
                # Save rode date from VFS
                buff = nodeFile.read(randomReadSize)
                # Close node file
                nodeFile.close()
                # Compare both data
                self.assertEqual(expectedContent, buff)
                # Reference file is automatically close ; 'with' statement
            checkAmount -= 1

    def test07_LargeSeek(self):
        """ #07 Large boundary seek
        Seek to limit of addressable offset, read and validate.
        First we create a huge file.

        Max :
        int32           2147483647              1.99 GB
        uint32          4294967295              3.99 GB
        ulonglong       18446744073709551615    16777215.99 TB
        
        """
        
        
    def runTest(self):
        pass


suite = unittest.TestSuite()
suite.addTest(LocalTests('test01_LoadOneFile'))
suite.addTest(LocalTests('test02_LsOneFile'))
suite.addTest(LocalTests('test03_LoadOneDirectory'))
suite.addTest(LocalTests('test04_ReadFile'))
suite.addTest(LocalTests('test05_RandomReadFile'))
res = unittest.TextTestRunner(verbosity=2).run(suite)

if (len(res.errors) or len(res.failures)):
    sys.exit(1)

#print "call -> LocalTests()"
#testC = LocalTests()
#print "call -> setUp()"
#testC.setUp()
#print "call -> test01_LoadOneFile()"
#testC.test01_LoadOneFile()
#print "call -> tearDown()"
#testC.tearDown()
